### DISCUSSION TITLE
WHy is it so slow?

### DISCUSSION THREAD

**Discussion by babakfp** [huntabyte/shadcn-svelte#425] [General]
Is it because of `lucide-svelte`? I installed `Button`, `DropdownMenu` and the reload speed went from x/ms to a few seconds :|

**Comment by ibanks42** [comment]
I noticed this as well and I do believe it's from lucide-svelte, but if you build it and run the production build it loads in quick/just fine. It seems to just be a dev thing.

What I ended up doing is just uninstalling lucide-svelte and making my own components from the SVGs when/where I needed them. A bit annoying, but oh well. If you go that route though there are quite a few components that have dependencies from radix-icons-svelte that you will have to fix as well.

**Comment by huntabyte** [comment]
Yeah it's likely the optimization of the icons. It should cache so once you deal with the slowness at first it'll speed up

**Comment by babakfp** [comment]
I think this speed issue results in a bad developer experience.

We could remove the index file and import the icon components individually; and add a note in docs letting people know why this is done and also that they can create their own index file (only for the icons that they use).

This solution may be making DX bad by asking devs to do more work, however I think the speed issue is worse and it can result it a worse DX.

Note on removing the index file, I think it should be removed because users may use it without knowing the cons and think that something is wrong with the library...

I think this issue needs to be resolved even with the caching feature and considering that the issue doesn't exist in production.

---

UPDATE:

Maybe it's a good idea to take a look into how other frameworks are solving this issue.

**Reply by huntabyte** [reply]
Does the same issue also occur when using `radix-icons-svelte` or other svelte icon libraries? I'm curious if this is an issue that could be resolved on Lucide's end?

**Reply by babakfp** [reply]
This isn't specific to Lucide. They import thousands/hundreds of icons in an index file, and when you just want to import a single icon, you kinda load them all, because you import the index file.

**Reply by babakfp** [reply]
https://github.com/huntabyte/shadcn-svelte/assets/44144724/de26ed25-83fc-448d-880b-8d4702a5465e

NOTE: in the `/import-all` route, I imported a single icon like `import { IconName } from "radix-icons-svelte"`.

**Comment by huntabyte** [comment]
See: https://github.com/lucide-icons/lucide/issues/1284

Right now I can't even import the component directly from lucide, which is an issue. Definitely not creating a bunch of SVGs either. if they don't enable us to resolve, then we'll switch to something like this: https://github.com/shinokada/svelte-lucide which appears to have accomplished what the original has not.

**Reply by leeoocca** [reply]
I'm not sure that PR solved the issue... It's been merged but loading still takes ages. I'm trying out `svelte-lucide` like you recommended, which seems to be working better.

**Comment by babakfp** [comment]
For the demo on the left side, I wrote a script that removed the import code for the unused icons from the index file (for both of the icon libraries).

https://github.com/huntabyte/shadcn-svelte/assets/44144724/20922a92-9216-46d1-bf12-7fa26f3c2e68

```
         FIRST LOAD    FULL LOAD
LEFT     16s           33s
RIGHT    26s           57s
```

**Reply by huntabyte** [reply]
Yes, but Lucide is the one that doesn't provide a way to directly import the component without svelte-check failing.

**Reply by hugotox** [reply]
I think they fixed this, I'm importing my icons like `import Camera from 'lucide-svelte/icons/camera'` and the svelte check passes just fine

**Comment by adaptive-shield-matrix** [comment]
Maybe use mdi icons instead?
https://pictogrammers.com/library/mdi/

**Comment by babakfp** [comment]
Can I create a library myself for lucide icons, and create a PR for replacing the libraries? I can do the same for melt-ui too since it suffers from the same issue too.

I made [this](https://www.npmjs.com/package/phosphor-icons-svelte) library for another icon library for the same reasons: 

- TypeScript support.
- Built with accessibility in mind.
- Improved performance and developer experience.

**Reply by huntabyte** [reply]
There's an active PR with lucide that will resolve this issue once it's merged https://github.com/lucide-icons/lucide/pull/1707

**Comment by armchair-traveller** [comment]
You guys not using https://github.com/unplugin/unplugin-icons ?

**Reply by FranckPoingt** [reply]
Came here to ask what would be the required effort to replace lucid with something like Iconify or unplugin. It would be great to have the option of choosing the icon library of our choice

**Reply by huntabyte** [reply]
We're in the discussion about a potential plugin-based system that could be integrated with the CLI to make these swaps, they'd be community-driven and maintained though.

**Reply by FranckPoingt** [reply]
@huntabyte great to hear, is this discussion public somewhere?

**Comment by cannap** [comment]
can i not just uninstall lucide or is it used internally?

**Reply by huntabyte** [reply]
It's not used internally so you can uninstall it and replace it with whatever you'd like!

**Comment by chientrm** [comment]
Uhm, Maybe replace `lucide-svelte` with `unplugin-icons` 😗

**Comment by hugotox** [comment]
Hello! 
FWIW I also thought the slowness issue was caused by icons, so as proof of concept I just removed all of them, and the slowness persisted but revealed another (similar) culprit: in dev mode, vite doesn't tree shake and it just transforms the whole library.

E.g on my page I'm just using the `Button`, 
import is the standard: 
```typescript
import { Button } from "$lib/components/ui/button";
``` 
Which makes the import from bits ui like this: 
```typescript
import { Button as ButtonPrimitive } from 'bits-ui'
```

That instructs vite to process the whole 3.2Mb lib, as shown:
<img width="683" alt="image" src="https://github.com/huntabyte/shadcn-svelte/assets/1758062/eaf6d252-b48d-47d2-acc8-87cb99ad0708">

Worth noting this issue doesn't happen in production builds, where the imports are correctly tree-shaken.

@huntabyte any chance to allow direct imports from bits-ui? So we can do something like 

```typescript
import ButtonPrimitive from 'bits-ui/button'`
```

**Reply by RonB** [reply]
Hi,
I am a bit confused and am not able to solve the enormous processing time when refreshing a page. If for instance a simple dialog or menu component is on the page then the page load time when refreshing is some 8 seconds or more. If I remove the component then the page is really fast. I am not using any icons.
When debugging vite I see it takes a long time when loading the first page and the server starts processing the bits-ui and melt-ui files. That is not a problem but at every subsequent page refresh vite again takes a long time to load the entry.js file generated in the .pnpm/@sveltjs+kit.... 
<img width="922" alt="image" src="https://github.com/huntabyte/shadcn-svelte/assets/559779/f41b280d-0b73-4964-b308-97f40d0b2e46">
Any help on this would be greatly appreciated.

**Reply by Masstronaut** [reply]
Have you also noticed they result in large bundles in prod? Not 3.2mb but not small.

**Reply by shyakadavis** [reply]
Hi;

I have not done any benchmarks, but I've been doing some reading on barrel files.

- https://dev.to/thepassle/a-practical-guide-against-barrel-files-for-library-authors-118c
- https://marvinh.dev/blog/speeding-up-javascript-ecosystem-part-7/
- e.t.c

Curious to know if this could also play a part in this?

**Comment by emmabyte** [comment]
One of the solutions I came up with is using svelte-fa for FontAwesome Icons and after adding a component to my project for shadcn-svelte, I replace the lucide icon with the corresponding one from svelte-fa. Easy fix without needing to manually create a bunch of different icons

**Reply by RonB** [reply]
Thanks for the tip but as I stripped out all of the icons during testing I don't think that is the problem. It might have something to do with the import *....  barrel import but I am not sure. Also, it does't make any difference wether the bits-ui and other depencies are in the depencies or the devDependencies section in package.json.

**Reply by emmabyte** [reply]
Hmm, okay you make a good point. Swapping out icons fixed build issues I had with Lucid, but the performance issue in dev is strange. I don't know enough about how Svelte/Vite do tree-shaking in dev mode, but that would definitely make sense to me. Curious to hear if anyone else comes up with something.

**Reply by hugotox** [reply]
This issue is definitely related to how vite loads lucide and bits-ui. Take a look at the bare bones example:
```
<script lang="ts">
  import { Button } from "$lib/components/ui/button/index.js";
</script>
 
<Button>Button</Button>
```
As you can see in the network tab, both libs are fully loaded, even when I'm not using icons on this example, lucide is still there:
<img width="1134" alt="Screenshot 2024-04-28 at 11 18 52 AM" src="https://github.com/huntabyte/shadcn-svelte/assets/1758062/62662fa2-3096-40b2-bfd1-51dc0962da20">

**Reply by RonB** [reply]
Ok thank you for your time. This is strange because this does not happen in my project. lucide-svelte is not loaded and not even present in my project. Bits-ui is 265kb so somehow our setups are different. And in your case bits-ui is loaded from the tooltip, and not from the button it seems.

![image](https://github.com/huntabyte/shadcn-svelte/assets/559779/121ee577-f288-4b07-8435-414fb87b00b9)

**Reply by emmabyte** [reply]
@RonB, that's interesting how your bits-ui is so small. Out of curiosity, what components are you using?

I'm seeing 3.6mb for bits-ui in my project, and I'm using the following shadcn-svelte components:
- accordion
- avatar
- button
- dropdown-menu
- input
- select
- sheet
- sonner

**Reply by RonB** [reply]
It does not matter how many components I load, it is always 265kb.
<img width="253" alt="image" src="https://github.com/huntabyte/shadcn-svelte/assets/559779/f90fe2b1-96b9-421e-93fb-4f9e36335b78">

**Reply by emmabyte** [reply]
Do you happen to have a public repo that others can clone to see if we also show the same size for bits-ui? My hunch is that the reason you aren't experiencing the issue is because of the dev environment you setup or some configuration setting you have set up with vite.

I love dev containers in VS Code, but I don't love the idea of being locked into using it just because of a vite issue.

**Reply by RonB** [reply]
About the dev container, It is more the cause of the problem than the solution. The access from the container to the sourcecode on the host seems to be the bottleneck, therefore I moved the sourcecode to a ramdisk to avoid the physical access to the ssd. So  you dont need a dev container at all, just make sure that file access to the source is fast. Than to you other request:

It is a private company repo so  it would take some time to create a public one, but I don't do anything special:

vite.config.ts
```
import { sveltekit } from '@sveltejs/kit/vite';
import { defineConfig } from 'vite';

export default defineConfig({
	server: {
		port: 3000,
		host: true
	},
	preview: {
		port: 4000
	},
	plugins: [
		sveltekit(),
	]
})
```

svelte.config.js
```
import adapter from "@sveltejs/adapter-cloudflare";
import preprocess from 'svelte-preprocess';
/** @type {import('@sveltejs/kit').Config} */
const config = {
	preprocess: [
		preprocess({
			postcss: true
		})
	],
	kit: {
		adapter: adapter(),
		csrf: {
			// handled in hooks.server.ts, because we can have multiple valid origins
			checkOrigin: false,
		},
		env: {
			dir: '.'
		},
		paths: {
			base: "",
		},
		alias: {
			$src: './src',
			$routes: './src/routes',
			$common: './src/common',
			$components: './src/lib/components',
			'$components/*': './src/lib/components/*',
		},
	},
};

export default config;
```
postcss.conig.js:
```
export default {
    plugins: {
        tailwindcss: {}
    }
};
```

tsconfig.json:
```
{
	"extends": "./.svelte-kit/tsconfig.json",
	"compilerOptions": {
		"lib": [
			"es5",
			"es6",
			"es7",
			"esnext",
			"dom"
		],
		"sourceMap": true,
		"importHelpers": true,
		"removeComments": false,
		"esModuleInterop": true,
		"moduleResolution": "node",
		"resolveJsonModule": true,
		"strict": true,
		"noUncheckedIndexedAccess": true,
		"skipLibCheck": true,
		"strictPropertyInitialization": false,
		"noUnusedLocals": true,
		"noUnusedParameters": true,
		"noImplicitReturns": true,
		"noImplicitThis": true,
		"noFallthroughCasesInSwitch": true,
		"downlevelIteration": true,
		"isolatedModules": true,
		"types": []
	}
}
```

pacakge.json
```
{
  "name": "www",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "dev": "vite dev",
    "build": "vite build",
    "preview": "vite preview",
    "check": "svelte-kit sync && svelte-check --tsconfig ./tsconfig.json",
    "check:watch": "svelte-kit sync && svelte-check --tsconfig ./tsconfig.json --watch",
    "lint": "prettier --check . && eslint .",
    "format": "prettier --write .",
    "pages:preview": "pnpm run build && wrangler pages dev --compatibility-date=2024-01-29 .svelte-kit/cloudflare",
    "pages:deploy": "pnpm run build && wrangler pages deploy .svelte-kit/cloudflare"
  },
  "devDependencies": {
    "@auth/core": "^0.28.2",
    "@auth/sveltekit": "^0.14.2",
    "@iconify/json": "^2.2.196",
    "@iconify/svelte": "^3.1.6",
    "@iconify/tools": "^4.0.2",
    "@playwright/test": "^1.42.1",
    "@prettier/sync": "0.3.0",
    "@sveltejs/adapter-cloudflare": "^4.4.0",
    "@sveltejs/kit": "^2.5.5",
    "@sveltejs/vite-plugin-svelte": "^3.0.2",
    "@types/cookie": "^0.6.0",
    "@types/node": "^20.12.2",
    "@types/swagger-ui": "^3.52.4",
    "@types/uuid": "^9.0.8",
    "@types/ws": "^8.5.10",
    "@typescript-eslint/eslint-plugin": "^7.4.0",
    "@typescript-eslint/parser": "^7.4.0",
    "@zodios/core": "^10.9.6",
    "@zodios/openapi": "^10.5.0",
    "autoprefixer": "^10.4.19",
    "cookie": "^0.6.0",
    "eslint": "^8.57.0",
    "eslint-config-prettier": "^9.1.0",
    "eslint-plugin-svelte": "^2.35.1",
    "mime": "^4.0.1",
    "mode-watcher": "^0.3.0",
    "paneforge": "^0.0.4",
    "postcss": "^8.4.38",
    "postcss-load-config": "^5.0.3",
    "svelte": "^4.2.12",
    "svelte-check": "^3.6.8",
    "svelte-confetti": "^1.3.2",
    "svelte-markdown": "^0.4.1",
    "svelte-preprocess": "^5.1.4",
    "tailwind-merge": "^2.3.0",
    "tailwind-variants": "^0.1.20",
    "tailwindcss": "^3.4.3",
    "ts-node": "^10.9.1",
    "tslib": "^2.6.2",
    "tsx": "^3.14.0",
    "typescript": "^5.4.3",
    "uuid": "^9.0.1",
    "vaul-svelte": "^0.3.0",
    "vite": "^5.2.10",
    "vite-node": "^1.5.2",
    "vite-plugin-tailwind-purgecss": "^0.3.3",
    "vitest": "^1.3.1",
    "ws": "^8.16.0",
    "zod-to-json-schema": "^3.22.4"
  },
  "dependencies": {
    "@internationalized/date": "^3.5.2",
    "@unovis/svelte": "1.3.6-beta.1",
    "@unovis/ts": "1.3.6-beta.1",
    "bits-ui": "^0.21.4",
    "clsx": "^2.1.1",
    "cmdk-sv": "^0.0.15",
    "d3-scale": "^4.0.2",
    "embla-carousel-autoplay": "8.0.0-rc19",
    "embla-carousel-svelte": "8.0.0-rc19",
    "formsnap": "^0.5.1",
    "highlight.js": "^11.9.0",
    "mode-watcher": "^0.2.2",
    "nanoid": "^5.0.6",
    "paneforge": "^0.0.2",
    "svelte-headless-table": "^0.18.2",
    "svelte-legos": "^0.2.2",
    "svelte-local-storage-store": "^0.5.0",
    "svelte-radix": "^1.1.0",
    "svelte-sonner": "^0.3.19",
    "sveltekit-superforms": "^2.11.0",
    "swagger-ui": "^5.11.0",
    "vaul-svelte": "^0.0.6",
    "zod": "^3.22.4"
  },
  "type": "module"
}
```

**Reply by emmabyte** [reply]
Thanks Ron, I'll try and reproduce with those config settings on a fresh repo today. Hopefully I can reproduce, otherwise we might be out of luck unless someone else can think of something.

**Comment by hugotox** [comment]
FWIW, deleting everything under `.svelte-kit` before starting the dev server made a huge improvement for me.
